proposals/infra/INF-0004-inproc-outofproc-compiler-library-support.md

```diff
@@ -0,0 +1,130 @@
+<!-- {% raw %} -->
+
+# Support for in-process, Out-of-process, or both for the c style compiler api
+
+* Proposal: [NNNN](INF-NNNN-inproc-outofproc-compiler-api-support.md)
+* Author(s): [Cooper Partin](https://github.com/coopp)
+* Sponsor: [Cooper Partin](https://github.com/coopp)
+* Status: **Under Consideration**
+* Impacted Project(s): (Clang)
+
+* Issues:
+
+## Introduction
+
+An effort is underway to bring compilation support into clang for HLSL based
+shaders.  A C style api will be built to enable applications to compile
+a shader from their own processes in addition to being able to launching the
+clang process.
+
+This document is to propose building only the out-of-process support which
+gives the caller a more performant and secure solution.
+
+### Why have an api at all?  Why not just run clang.exe yourself?
+
+Before diving into any details we should discuss why we even need an
+api at all.  Why can't clients just execute the clang.exe with options
+and compile their shaders that way?  They can and they will. Launching clang
+to compile a shader will be supported.
+
+A design requirement that cannot be satisfied with just executing clang with a 
+command-line is _include handlers_. An include handler is a way for clients to
+hook into the compilation process and dynamically load dependencies when
+needed. We have already been asked by a game studio to support this feature.
+This can only be achieved if there is an api to support adding these handlers.
+
+### Overview of api hosting differences
+
+Here are two common ways an api can be loaded and used.
+
+* **In Process** - An application calls an api that loads compilation support
+into the application's process. All work is performed within the application's
+process.
+
+* **Out of Process** - An application calls an api that launches one or more
+processes to perform the compilation. All work is performed outside the
+calling application's process.
+
+## Motivation
+
+### Why prefer an out of process solution?
+
+* Process creation on Windows based systems is expensive.  Launching clang.exe
+over and over again to compile shaders becomes a lengthy operation on 
+Windows platforms.
```

Comment:
> Launching clang.exe over and over again to compile shaders....

This is actually describing one possible out-of-proc solution.  In-proc would not involve launching new processes.

Reply:
I removed this.  You are correct that it is referring to a solution and that isn't want the list is supposed to communicate.